lib/api/api.js

```diff
@@ -184,14 +186,115 @@ const api = {
             }
             return { returnTagCount, isImplicitDeny };
         }
+        let formDataParser;
+        let fileEventData = null;
+
+        if (apiMethod === 'objectPost') {
+            if (request.headers['content-type'].includes('multipart/form-data')) {
+                formDataParser = busboy({ headers: request.headers });
+            } else {
+                const contentTypeError = errors.PreconditionFailed
+                    .customizeDescription('Bucket POST must be of the enclosure-type multipart/form-data');
+                return process.nextTick(callback, contentTypeError);
+            }
+        }
 
         return async.waterfall([
+            next => {
+                if (apiMethod === 'objectPut' || apiMethod === 'objectPutPart') {
+                    return next(null);
+                }
+                if (apiMethod === 'objectPost' && request.headers['content-type'].includes('multipart/form-data')) {
+                    writeContinue(request, response);
+
+                    request.formData = {};
+                    let totalFieldSize = 0;
+                    const MAX_FIELD_SIZE = 20 * 1024; // 20KB
+                    formDataParser.on('field', (fieldname, val) => {
+                        totalFieldSize += Buffer.byteLength(val, 'utf8');
+                        if (totalFieldSize > MAX_FIELD_SIZE) {
+                            const err = errors.MaxPostPreDataLengthExceeded;
+                            formDataParser.emit('error', err);
+                        }
+                        // Convert fieldname to lowercase for case-insensitive comparison
+                        const lowerFieldname = fieldname.toLowerCase();
+                        request.formData[lowerFieldname] = val;
+                    });
+
+                    formDataParser.on('file', (fieldname, file, filename, encoding, mimetype) => {
+                        fileEventData = { fieldname, file, filename, encoding, mimetype };
+                        // no key field == error
+                        if ('key' in request.formData) {
+                            return next(null);
```

Comment:
It should only be the field `file`  that should be the data to save.

If field `key` is a file that trigger a field event the content of the file should be the key.

And if `file` field is a string it should be managed in the `field` event

Reply:
As this is in the 'file' event, we need to start the process of streaming the file. If it's a large file we need to start this process by calling `next` before 'finish' will be emitted. Am I misunderstanding?

Reply:
Actually it should not be called in the `finish` event, the `finish` is triggered once all the file stream have been read

Reply:
To be clear, this condition for key is to check if a key field is found before processing the file.
As file is meant to be last, if there is not a key found before that, it is an error.
This is as AWS behaves, ignoring fields after the file.

Reply:
The `file` field needs to be last but if it is text type and not file type it will trigger the `field` event and not the `file` event.

If any other field like `key` is received as file type, you should not stop parsing

Reply:
Unless I misunderstand, this has been resolved now with the updates I made to the whole flow.

Reply:
see https://github.com/scality/cloudserver/pull/5601#discussion_r1690995359 I'm closing this